controllers/profiles/reconciler.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2023 Red Hat, Inc. and/or its affiliates
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package profiles
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	operatorapi "github.com/kiegroup/kogito-serverless-operator/api/v1alpha08"
+)
+
+// ProfileReconciler is the public interface to have access to this package and perform the actual reconciliation flow.
+//
+// There are a few concepts in this package that you need to understand before attempting to maintain it:
+//
+// 1. ProfileReconciler: it's the main interface that internal structs implement via the baseReconciler.
+// Every profile must embed the baseReconciler.
+//
+// 2. stateSupport: is a struct with a few support objects passed around the reconciliation states like the client and logger.
+//
+// 3. reconciliationStateMachine: is a struct within the ProfileReconciler that do the actual reconciliation.
+// Each part of the reconciliation algorithm is a ReconciliationState that will be executed based on the ReconciliationState.CanReconcile call.
+//
+// 4. ReconciliationState: is where your business code should be focused on. Each state should react to a specific operatorapi.ConditionType.
+// The last conditions your state handles, the better. It can provide specific code that will only be triggered if the workflow is in that specific condition.
+//
+// 5. objectCreator: are functions to create a specific Kubernetes object based on a given workflow instance. This function should return the desired default state.
+//
+// 6. mutateVisitor: is a function that states can pass to objectEnsurer that will be applied to a given live object during the reconciliation cycle.
+// For example, if you wish to guarantee that an image in a specific container in the Deployment that you control and own won't change, make sure that your
+// mutate function guarantees that.
```

Comment:
isn't better control it with annotation or label?

Reply:
Not sure if I understood. This is the logic behind the reconciliation which one should pass a callback to mutate the object to the desired state.